modules/src/ics04_channel/handler/chan_open_init.rs

```diff
@@ -0,0 +1,201 @@
+
+        let context = MockContext::default();
+
+        let msg_conn_init =
+            MsgConnectionOpenInit::try_from(get_dummy_msg_conn_open_init()).unwrap();
+
+        let init_conn_end = ConnectionEnd::new(
+            ConnectionState::Init,
+            msg_conn_init.client_id().clone(),
+            msg_conn_init.counterparty().clone(),
+            get_compatible_versions(),
+            msg_conn_init.delay_period,
+        );
+
+        let ccid = <ConnectionId as FromStr>::from_str("defaultConnection-0");
+        let cid = match ccid {
+            Ok(v) => v,
+            Err(_e) => ConnectionId::default(),
+        };
+
+        let tests: Vec<Test> = vec![
+            Test {
+                name: "Processing fails because no connection exists in the context".to_string(),
+                ctx: context.clone(),
+                msg: ChannelMsg::ChannelOpenInit(msg_chan_init.clone()),
+                want_pass: false,
+            },
+            Test {
+                name: "Processing fails because port does not have a capability associated"
+                    .to_string(),
+                ctx: context
+                    .clone()
+                    .with_connection(cid.clone(), init_conn_end.clone()),
+                msg: ChannelMsg::ChannelOpenInit(msg_chan_init.clone()),
+                want_pass: false,
+            },
+            Test {
+                name: "Good parameters".to_string(),
+                ctx: context.with_connection_and_port_capability(
+                    MsgChannelOpenInit::try_from(get_dummy_raw_msg_chan_open_init())
+                        .unwrap()
+                        .port_id()
+                        .clone(),
+                    cid,
+                    init_conn_end,
+                ),
+                msg: ChannelMsg::ChannelOpenInit(msg_chan_init),
+                want_pass: true,
+            },
+        ]
+        .into_iter()
+        .collect();
+
+        for test in tests {
+            let res = dispatch(&test.ctx, test.msg.clone());
+            // Additionally check the events and the output objects in the result.
+            match res {
+                Ok(proto_output) => {
+                    assert_eq!(
+                        test.want_pass,
+                        true,
+                        "conn_open_init: test passed but was supposed to fail for test: {}, \nparams {:?} {:?}",
+                        test.name,
+                        test.msg.clone(),
+                        test.ctx.clone()
+                    );
+                    assert_ne!(proto_output.events.is_empty(), true); // Some events must exist.
+
+                    // The object in the output is a ConnectionEnd, should have init state.
+                    let res: ChannelResult = proto_output.result;
+                    //assert_eq!(res.channel_id, msg_chan_init.channel_id().clone());
```

Comment:
Could we have a test checking the port_id? Something like:

```suggestion
                    assert_eq!(res.port_id, msg_chan_init.port_id);
```

(Watch out for compilation issues, a `clone` may be necessary to `msg_chan_init`.)